backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/predicate/impl/ElasticsearchKnnPredicate.java

```diff
@@ -0,0 +1,261 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.backend.elasticsearch.search.predicate.impl;
+
+import java.lang.invoke.MethodHandles;
+import java.lang.reflect.Array;
+
+import org.hibernate.search.backend.elasticsearch.gson.impl.JsonAccessor;
+import org.hibernate.search.backend.elasticsearch.gson.impl.JsonArrayAccessor;
+import org.hibernate.search.backend.elasticsearch.gson.impl.JsonObjectAccessor;
+import org.hibernate.search.backend.elasticsearch.logging.impl.Log;
+import org.hibernate.search.backend.elasticsearch.search.common.impl.AbstractElasticsearchCodecAwareSearchQueryElementFactory;
+import org.hibernate.search.backend.elasticsearch.search.common.impl.ElasticsearchSearchIndexScope;
+import org.hibernate.search.backend.elasticsearch.search.common.impl.ElasticsearchSearchIndexValueFieldContext;
+import org.hibernate.search.backend.elasticsearch.types.codec.impl.ElasticsearchFieldCodec;
+import org.hibernate.search.backend.elasticsearch.types.codec.impl.ElasticsearchVectorFieldCodec;
+import org.hibernate.search.engine.search.predicate.SearchPredicate;
+import org.hibernate.search.engine.search.predicate.spi.KnnPredicateBuilder;
+import org.hibernate.search.util.common.AssertionFailure;
+import org.hibernate.search.util.common.logging.impl.LoggerFactory;
+
+import com.google.gson.JsonArray;
+import com.google.gson.JsonObject;
+
+public abstract class ElasticsearchKnnPredicate extends AbstractElasticsearchSingleFieldPredicate {
+
+	private static final Log log = LoggerFactory.make( Log.class, MethodHandles.lookup() );
+
+	protected final ElasticsearchSearchPredicate filter;
+	protected final int k;
+	protected final JsonArray vector;
+
+	private ElasticsearchKnnPredicate(AbstractKnnBuilder<?> builder) {
+		super( builder );
+		this.filter = builder.filter;
```

Comment:
Dumb question maybe but... what happens if the filter itself is a KNN predicate?

Reply:
I've tried 
```json
.where( f -> f.knn( 3 )
	.field( "location" )
	.matching( 5f, 4f )
	.filter(
			f.knn( 6 )
					.field( "location" )
					.matching( 6f, 3f )
	)
```
And it seems to work correctly, it "finds 6 close to 6-3 and then finds 3 close to 5-4 among those 6 ones...

Reply:
Do I understand correctly it returns 6 documents?

I'd expect this query to only return 3 documents, since the top-level one has k=3.

And if it does return 3 documents with your current implementation... does that mean Elasticsearch executes a conjunction of all knn predicates? If so, we need to reconsider how we require knn predicates to be structured in the Search DSL.

Reply:
no no, I should've posted a complete example: https://github.com/hibernate/hibernate-search/pull/3871/files#diff-03164f33f7d3eedeb9d14e03be09490770e107dfe30661215c26dd3c3ebd857cR648-R667
what I meant was that the filter removes some elements and then the top knn finds 3 "matches"  (among those that inner knn kept). so the result is 3 elements. 
For Elasticsearch ... If I'd execute:
```json
{
  "knn": [
    {
      "field": "location",
      "k": 6,
      "num_candidates": 6,
      "query_vector": [
        6.0,
        3.0
      ]
    },
    {
      "field": "location",
      "k": 3,
      "num_candidates": 3,
      "query_vector": [
        5.0,
        4.0
      ]
    }
  ],
  "_source": false
}
```
I'd get 6 docs: `["ID:2", "ID:3", "ID:1", "ID:6", "ID:12", "ID:4"]`, so it's a disjunction for Elasticsearch. And it seems "fine" the way we are allowing them to be added as should clauses of a top level bool. 
As for this particular scenario I'd say  we just need to not allow a knn as a filter if it's an Elasticsearch distribution (which should be doable with that other comment on predicate context in mind 😃 🙈 )

Reply:
> And it seems "fine" the way we are allowing them to be added as should clauses of a top level bool.

To be clear the current behavior on Elasticsearch is definitely not fine, because the results are incorrect. There should only be 3.

It would be fine if that case was simply forbidden in Elasticsearch, because then users wouldn't experience this (incorrect) result.

>  As for this particular scenario I'd say we just need to not allow a knn as a filter if it's an Elasticsearch distribution (which should be doable with that other comment on predicate context in mind 😃 🙈 )

Agreed. A "filter" clause is not a "should" clause, after all. This would only be consistent with other constraints in boolean predicates.

Reply:
Sorry 🙈 so the example with Elasticsearch is just adding two knn predicates (if we want to do that in Search it's through two should clauses). if I'd try to actually put knn as a filter for Elasticsearch it's an exception. 
I was just saying that adding multiple knn clauses for Elasticsearch is a disjunction